packages/nimble-components/src/table/specs/README.md

```diff
@@ -132,13 +132,17 @@ The various APIs/features of the `nimble-table` will be split up amongst several
     -   Define how the action menu gets associated with a particular column
     -   Define the table-level(column-level?) API(s) for applying an action menu to the table (slot, properties, etc...)
 
+_Attributes_
+
+-   `id-field-name` - An optional string attribute that specifies the field name within a row's record to use as a row's ID. If the attribute is not defined or a field with the specified name does not exist in a row's record, a default ID will be created. The default ID is the index of the row's record within the table's `data` property. The value associated with the field name must be unique across all records in the table's data.
```

Comment:
> The default ID is the index of the row's record within the table's `data` property. The value associated with the field name must be unique across all records in the table's data.

Will there be any assertions for these? What if one record is missing the id but the rest have it? What if an id is not unique? Does tanstack give us any defined behavior there? (sounds like we should have tests too)

Reply:
These are similar questions to what @jattasNI asked in a comment above. Let's consolidate the discussion to this thread.

The current behavior we get for free from TanStack is this:
- There are no assertions for uniqueness within TanStack. You'd just have multiple rows with the same ID, which could cause indeterministic behavior when applying state (like row selection) or when firing an event to tell the user what row an action occurred on. There is a function in TanStack called `getRow(id: string): Row`, which would only return one row even if more than one with a given ID was found -- from the code I've seen, I think it would always return the last row in the table with the given ID.
- There is a hook in TanStack to implement `getRowId(row: TData, index: number, parent?: Row<TData>): string` to return whatever ID we want. If we don't implement that function, or if our implementation of that function returns a falsey value, then the default index-based ID will be used by TanStack.
- We aren't in control of when `getRowId()` is called, so the options I can see for providing our own assertions about the ID being present & unique are:
    1. Check every time the data is set that our `getRowId()` function will return a unique value for every record. This means that `getRowId()` would be called twice every time the data is updated, but maybe that isn't too bad since it should be a very light-weight function.
    2. Override `getCoreRowModel` from TanStack. This is the function that calls `getRowId()` and builds a cache of Rows and a map from id to Row. It would be easy to add a check here that the ID was unique, but I don't like the idea of maintaining our own copy of the implementation of `getCoreRowModel`. While it isn't too complex, it's large enough that I'd rather not duplicate it.
    3. Try to upstream a change to TanStack that performs the assertion within their implementation of `getCoreRowModel`.

Given those options, I'd lean towards option 1, but I'd be interested in hearing other opinions or ideas.

Reply:
iv. Don't do any automatic validation in production code. We could provide a helper function to do validation and clients could decide when to call it. We could recommend that they call it in tests after setting the data and maybe also in local dev builds. (A variation of this would be to have strict validation be a mode of the table rather than a function to be called, but still recommend only enabling strict validation in non-production scenarios.

Something like this is where I'm leaning initially (despite being one of the people to ask about automating assertions 😛) since I don't think it will be common to accidentally generate non-unique IDs so I'm wary of adding runtime cost to assert it.

Reply:
I also lean (_hard_) towards Jesse's "iv" option, particularly with having the strict validation be a mode on the table. Note that I also think this could be done as follow-on work to introducing the attribute on the table.

Reply:
I'd kinda lean the other way. Be strict by default and give the option to loosen up / shoot yourself in the foot (maybe an attribute like `skip-load-validation` or something if we find it is needed).

Since the validation will only happen on data updates and not interactions / scrolling, I think it's not going to generally be a problem (particularly for our target size of ~10K elements). We did similar order expensive things on grid data load in WebVIs (full data copy >.<) and found the perf to be pretty good.

Reply:
I personally don't think that many (or maybe any) clients would call a helper function to do validation. At least in SLE, tests are all using fake data, so that validation doesn't help much within tests -- other than potentially telling them that their mock data is invalid.

Therefore, I think we should either always do validation or never do validation. I personally don't think we'll find that the validation step is that expensive because all we're really doing is accessing a single property within each data item, and this would always be done when we're about to create a whole new collection of TanStack Row models for each data item (which is a significantly more expensive operation).

Reply:
I love that our initial perspectives cover pretty much the entire spectrum! Let's talk about it at/after standup today to reach consensus.

Reply:
I've updated the spec as follows:
- we'll always do validation
- if validation fails, no rows will be rendered in the table
- the table will report that it is invalid through a `validity` property and a `checkValidity()` function